### ISSUE TITLE
GetGradient in PetscNonlinearSolver and JFNK in general

### ISSUE THREAD

**Issue by tangqi** [mfem/mfem#906]
I am looking at petsc/ex10p. I have two questions. 

1. Is it true `GetGradient` in `__mfem_snes_jacobian` is only called once in ex10p when the snes solver is used? So Jacobian is magically linked with the updated solution every time step? I asked because `NewtonSolver::Mult` call `GetGradient` whenever the solution is updated. But `PetscNonlinearSolver::Mult` does not explicitly update Jacobian, unless I miss something.

2. I need a matrix-free Newton-Krylov with a "physical-based" preconditioner. By changing petsc.cpp slightly, `-snes_mf_operator` is working well for me. I believe petsc would get the job done. Is there any other version of JFNK available? I know KINSOL is available but I do not have experience in adding my own preconditioner there. How hard would it be to add my own block-matrix preconditioner and link some blocks with Hypre? Any recommendation between KINSOL vs petsc for my specific goal?

Thanks,
Qi

**Comment by tzanio** [comment]
@stefanozampini, can you help Qi?

**Comment by stefanozampini** [comment]
__mfem_snes_jacobian is a callback that PETSc calls any time it needs a new Jacobian evaluation.
PetscNonlinearSolver::Mult basically calls SNESSolve() from PETSc. The default SNES method is Newton with line search; at each Newton iteration, PETSc calls SNESComputeJacobian, which in turn calls __mfem_snes_jacobian that calls the user GetGradient method

**Comment by stefanozampini** [comment]
Regarding point number 2. What do you mean by changing petsc.cpp slightly? What version of MFEM are you using? what didn't work for you?
You can use a PetscPreconditionerFactory in MFEM to achieve physics based preconditioning.
If you tell me what you want to achieve, I can add a case to ex10p to show you how to do it.

**Comment by tangqi** [comment]
@stefanozampini Thanks for your quick reply. Q1 makes sense. 

I want to do matrix-free Newton-Krylov with my own preconditioner, starting with the flag of `-snes_mf_operator`. But when I tested it in petsc/ex10.cpp, it complains Mat A and P in `__mfem_snes_jacobian` is not assembled (in this case, it will be used as a preconditioner in JFNK). So I changed it slightly by adding
```c++
   // Avoid unneeded copy of the matrix by hacking
   Mat B = pA->ReleaseMat(false);
   ierr = MatHeaderReplace(P,&B); CHKERRQ(ierr);

   MatAssemblyBegin(P,MAT_FINAL_ASSEMBLY);
   MatAssemblyEnd(P,MAT_FINAL_ASSEMBLY);
   if (A!=P)
   {
        MatAssemblyBegin(A,MAT_FINAL_ASSEMBLY);
        MatAssemblyEnd(A,MAT_FINAL_ASSEMBLY);
   }
```
Was pA of PetscParMatrix not assembled in the original `__mfem_snes_jacobian`?

**Comment by tangqi** [comment]
Sorry @stefanozampini , I missed your good suggestion on `PetscPreconditionerFactory`. That is actually close to what I have in mind. Are you aware of any good examples on how to use in mfem or similar examples in petsc?

I am using the newest petsc and have merged mfem master to my branch. Thanks again!

**Comment by stefanozampini** [comment]
I will fix the issue with snes_mf_operator, thanks for reporting it. And I'll add an example that uses the factory. @tzanio when are you planning to release?

**Comment by tzanio** [comment]
> when are you planning to release?

May 22nd

**Comment by tangqi** [comment]
Thanks a lot @stefanozampini

I have read pcshell in petsc.cpp. I have two questions but I think I am very close to what I needed (apologize for many questions).

1. If I understand correctly, I need to define NewPreconditioner in PetscPreconditionerFactory, which will be used as op->Mult(xx,yy) in __mfem_pc_shell_apply. Is there any example on how to define Mult that has the input of PetscParVector? If not, I will wait for your new example.

2. Following Q1, when defining Mult that uses input of PetscParVector in pcshell, is it better to convert things back to Vector and use operators of HypreParMatrix, or should I convert all HypreParMatrix to PetscParMatrix and then apply them to PetscParVector? For the first option, I would wrap the input Vecs to mfem::Vector (I assume I would recover the original block order and partitioning. Would it be possible?), then apply the working HypreParMatrix operators to those wrapped vectors. Since I need BoomerAMG anyway, that looks like a better option for me if it works.

If you can make some comments, that would be helpful. I could start to implement a few things before seeing your example.

Thanks,
Qi

**Comment by stefanozampini** [comment]
I have just updated the pull request https://github.com/mfem/mfem/pull/756 with a fix for -snes_mf_operator 
 https://github.com/mfem/mfem/pull/756/commits/69981d62ebdc8178656f49102c39acf894f2fbf9 
I have also added an example of usage and tests here https://github.com/mfem/mfem/pull/756/commits/c5586d5e5a7095ffbfc0317fbcef2961bb2a2622

Tzanio, why you still have to merge the PR?

**Comment by stefanozampini** [comment]
> If I understand correctly, I need to define NewPreconditioner in PetscPreconditionerFactory, which will be used as op->Mult(xx,yy) in __mfem_pc_shell_apply. Is there any example on how to define Mult that has the input of PetscParVector? If not, I will wait for your new example.

if you take a look at static PetscErrorCode __mfem_pc_shell_apply(PC pc, Vec x, Vec y), the Mult method is fed with PetscParVectors

**Comment by stefanozampini** [comment]
I don't understand Q2. If you do not need the PETSc AIJ format inside the newton solver, you can use PetscSolver::SetJacobianType(Operator::PETSC_MATHYPRE) and PETSc will use the HypreParCSR format, provided you have configured PETSc with support for HYPRE (either via --download-hypre or --with-hypre-dir=prefix_location_for_hypre).

With -snes_mf_operator, the OperatorHandle in NewPreconditioner will contain a PetscParMatrix wrapper of the HypreParCSR matrix constructed in the GetGradient operator. You can then do something like this in NewPreconditioner  

```
PetscParMatrix *pA;
oh.Get(pA);
hypre_ParCSRMatrix *parcsr;
MatHYPREGetParCSR(*pA,&parcsr);
```

**Comment by tangqi** [comment]
@stefanozampini Let me be more specific. (Thanks for suggestions on use HypreParCSR in petsc. I did not know that.)

I want to solve J du = -F(u) in a matrix-free fashion (inexact Newton) that needs a right physics-based preconditioner 
<a href="https://www.codecogs.com/eqnedit.php?latex=P&space;\begin{bmatrix}&space;x_1&space;\\&space;x_2&space;\\&space;x_3&space;\end{bmatrix}&space;=&space;\begin{bmatrix}&space;y_1&space;\\&space;y_2&space;\\&space;y_3&space;\end{bmatrix}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?P&space;\begin{bmatrix}&space;x_1&space;\\&space;x_2&space;\\&space;x_3&space;\end{bmatrix}&space;=&space;\begin{bmatrix}&space;y_1&space;\\&space;y_2&space;\\&space;y_3&space;\end{bmatrix}" title="P \begin{bmatrix} x_1 \\ x_2 \\ x_3 \end{bmatrix} = \begin{bmatrix} y_1 \\ y_2 \\ y_3 \end{bmatrix}" /></a>
My preconditioner looks like this
<a href="https://www.codecogs.com/eqnedit.php?latex=\begin{bmatrix}&space;K&space;&&space;0&space;&&space;M&space;\\&space;0&space;&&space;S&space;&&space;0&space;\\&space;A&space;&&space;N&space;&&space;0&space;\end{bmatrix}&space;\begin{bmatrix}&space;x_1&space;\\&space;x_2&space;\\&space;x_3&space;\end{bmatrix}&space;=&space;\begin{bmatrix}&space;y_1&space;\\&space;y_2&space;\\&space;K^{-1}&space;y_3&space;\end{bmatrix}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\begin{bmatrix}&space;K&space;&&space;0&space;&&space;M&space;\\&space;0&space;&&space;S&space;&&space;0&space;\\&space;A&space;&&space;N&space;&&space;0&space;\end{bmatrix}&space;\begin{bmatrix}&space;x_1&space;\\&space;x_2&space;\\&space;x_3&space;\end{bmatrix}&space;=&space;\begin{bmatrix}&space;y_1&space;\\&space;y_2&space;\\&space;K^{-1}&space;y_3&space;\end{bmatrix}" title="\begin{bmatrix} K & 0 & M \\ 0 & S & 0 \\ A & N & 0 \end{bmatrix} \begin{bmatrix} x_1 \\ x_2 \\ x_3 \end{bmatrix} = \begin{bmatrix} y_1 \\ y_2 \\ K^{-1} y_3 \end{bmatrix}" /></a>
Here S is a schur complement. Presumably, I have linked it with petsc and know how to implement preconditioner in mfem  (using HypreParMatrix and invert some operators using boomeramg). 

So the question is how to implement it using pcshell. I hope the implementation would be generic enough, as we will look at a more complicated system soon. The operator has to be inverted in a semi-implicit fashion, so I do not know if fieldsplit would be useful. 

Previously, we have implemented the whole preconditioner in pcshell completely by ourself using geometric multigrid and so on (in another code). But it sounds like there should be a better way. 

If the discussion has been too technical for the board, I can send you an email to discuss if you do not mind. 

Thanks a lot for the help!
Qi

**Comment by stefanozampini** [comment]
It looks like you need to solve 3 separate linear systems (1 for each block) in sequence.

Note that, if you return an MFEM block operator in the GetGradient() method, and you use MATHYPRE as Jacobian type for the nonlinear solver, then the operator handle will contain a PetscParMatrix that uses the PETSc MATNEST https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Mat/MATNEST.html format. Each of the nonzero block will be a MATHYPRE, for which you can access the ParCSR pointer as I have shown you before.

If you are explicitly constructing your Schur complement in the GetGradient method, then it is easy to write the preconditioner application. If not, you can create a matrix that behaves like a Schur complement https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/KSP/MatCreateSchurComplement.html#MatCreateSchurComplement, create an approximate preconditioner, and then feed everything yo a KSP

**Comment by stefanozampini** [comment]
Feel free to send me an email

**Comment by tangqi** [comment]
@stefanozampini Thanks again. I think I know how to do it after glancing through your branch and comments. 

I could implement my own GetGradient which is a block operator (Yes, I will build S by myself) of
<a href="https://www.codecogs.com/eqnedit.php?latex=\begin{bmatrix}&space;K&space;&&space;0&space;&&space;M&space;\\&space;0&space;&&space;S&space;&&space;0&space;\\&space;A&space;&&space;N&space;&&space;0&space;\end{bmatrix}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\begin{bmatrix}&space;K&space;&&space;0&space;&&space;M&space;\\&space;0&space;&&space;S&space;&&space;0&space;\\&space;A&space;&&space;N&space;&&space;0&space;\end{bmatrix}" title="\begin{bmatrix} K & 0 & M \\ 0 & S & 0 \\ A & N & 0 \end{bmatrix}" /></a>
Based on your comments, I will have access to the pointer of each nonzero block. So in the pc_shell_apply, I only need to call several KSPSolve in sequence using those block matrices. 

In other words, the main thing I need to implement is defining my own GetGradient and a special __mfem_pc_shell_apply. (I assume I could ignore apply_transpose if using GMRES.) Does that make sense?

**Comment by tangqi** [comment]
The question is answered through private communication. I will close this issue for now. Thanks again for the help!

**Comment by tangqi** [comment]
Hi @stefanozampini,

Attached is my simplified pcshell preconditioner based on your suggestion:
```c++
class MyBlockSolver : public Solver
{
private:
   Mat **sub;
   KSP kspblock[3];  // Create internal KSP objects to handle the subproblems
   mutable PetscParVector *X, *Y;  // Create PetscParVectors as placeholders X and Y
   IS index_set[3];
public:
   MyBlockSolver(const OperatorHandle &oh);
   virtual void SetOperator (const Operator &op);
   virtual void Mult(const Vector &x, Vector &y) const;
   virtual ~MyBlockSolver();
};
MyBlockSolver::MyBlockSolver(const OperatorHandle &oh) : Solver() { 
   // Get the PetscParMatrix out of oh.       
   PetscParMatrix *PP;
   oh.Get(PP);
   Mat P = *PP; // type cast to Petsc Mat
   PetscInt M, N;
   ierr=MatNestGetSubMats(P,&N,&M,&sub); // sub is an N by M array of matrices
   ierr=MatNestGetISs(P, index_set, NULL);  // get the index sets of the blocks

   X = new PetscParVector(P, true, false); 
   Y = new PetscParVector(P, false, false);

   for (int i=0; i<3; i++)
   {
     KSPCreate(PETSC_COMM_WORLD, &kspblock[i]);
     KSPSetOperators(kspblock[i], sub[i][i], sub[i][i]);
     KSPSetFromOptions(kspblock[i]);
     KSPSetUp(kspblock[i]);
   }
}
```
Do you see any obvious mistake? Is it okay to use new PetscParVector when P is MATNEST?

Here my preconditioner from GetGradient is a BlockOperator and I did not set my jacobian type. As a result, do I have pointers to all the sub matrices of type MATAIJ (in Mat **sub)?

Thanks again,
Qi
